alvr/client_openxr/src/lib.rs

```diff
@@ -723,47 +723,38 @@ pub fn entry_point() {
                     thread::yield_now();
                 }
 
-                if let Some((timestamp, hardware_buffer)) = frame_result {
-                    let mut history_views = None;
-                    for history_frame in &*views_history.lock() {
-                        if history_frame.timestamp == timestamp {
-                            history_views = Some(history_frame.views.clone());
-                        }
-                    }
-
-                    views = if let Some(views) = history_views {
-                        last_swapchain_left_view.insert(left_swapchain_idx, views[0]);
-                        last_swapchain_right_view.insert(right_swapchain_idx, views[1]);
-
-                        views
-                    } else {
-                        vec![default_view, default_view]
-                    };
-
-                    alvr_client_core::opengl::render_stream(
-                        hardware_buffer,
-                        [left_swapchain_idx, right_swapchain_idx],
-                    );
+                let (timestamp, hardware_buffer) = if let Some(pair) = frame_result {
+                    pair
+                } else {
+                    warn!("Timed out when waiting for frame!");
+                    (vsync_time, ptr::null_mut())
+                };
 
-                    if let Some(now) = xr_runtime_now(&xr_instance, platform) {
-                        alvr_client_core::report_submit(timestamp, vsync_time.saturating_sub(now));
+                let mut history_views = None;
+                for history_frame in &*views_history.lock() {
+                    if history_frame.timestamp == timestamp {
+                        history_views = Some(history_frame.views.clone());
                     }
+                }
 
-                    display_time = timestamp;
+                views = if let Some(views) = history_views {
+                    last_good_views = views.clone();
+                    views
                 } else {
-                    warn!("Timed out when waiting for frame!");
-                    views = if let (Some(left_view), Some(right_view)) = (
-                        last_swapchain_left_view.get(&left_swapchain_idx),
-                        last_swapchain_right_view.get(&right_swapchain_idx),
-                    ) {
-                        vec![*left_view, *right_view]
-                    } else {
-                        vec![default_view, default_view]
-                    };
-
-                    display_time = vsync_time;
+                    last_good_views.clone()
+                };
+
+                alvr_client_core::opengl::render_stream(
+                    hardware_buffer,
+                    [left_swapchain_idx, right_swapchain_idx],
+                );
+
+                if let Some(now) = xr_runtime_now(&xr_instance, platform) {
+                    alvr_client_core::report_submit(timestamp, vsync_time.saturating_sub(now));
                 }
```

Comment:
As @deiteris suggested, this should be invoked only when a new frame is computed.